### PR TITLE
Add labels to HPA and VPA from their their templates

### PR DESCRIPTION
**Pull request by ggaurav10**
**What this PR does / why we need it**:
Add labels to HPA and VPA from their their templates

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

File: controllers/hvpa_controller.go [hunk 1]
```diff
@@ -274,14 +274,22 @@ func (r *HvpaReconciler) deleteHvpaFinalizers(hvpa *autoscalingv1alpha1.Hvpa) {
...
+	}
+
 	// Updater policy set to "Off", as we don't want vpa-updater to act on recommendations
 	updatePolicy := vpa_api.UpdateModeOff
 
 	return &vpa_api.VerticalPodAutoscaler{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      hvpa.Name + "-vpa",
 			Namespace: hvpa.Namespace,
+			Labels:    labels,
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: What about annotations? Why not copy `hvpa.Spec.Vpa.Template.ObjectMeta` and override only `Name` and `Namespace`?

File: controllers/hvpa_controller.go [hunk 2]
```diff
@@ -340,10 +351,18 @@ func (r *HvpaReconciler) reconcileVpa(hvpa *autoscalingv1alpha1.Hvpa) (*vpa_api.
...
+	if labels == nil || len(labels) == 0 {
+		// TODO: Could be done better as part of validation
+		return nil, fmt.Errorf("Need labels in HPA template")
+	}
+
 	hpa := &autoscaling.HorizontalPodAutoscaler{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      hvpa.Name + "-hpa",
 			Namespace: hvpa.Namespace,
+			Labels:    labels,
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: Same as [above](https://github.com/gardener/hvpa-controller/pull/26/files#diff-6354ac486a904d10093c497f09f83a3eR292).

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Can we not copy the `ObjectMeta` from the templates and just override `Name` and `Namespace`?

File: controllers/hvpa_controller.go (hunk 1 above)
Comment (ggaurav10) [reply controllers/hvpa_controller.go]: Fixed [de8e35e](https://github.com/gardener/hvpa-controller/pull/26/commits/de8e35e64ae86f048cc812d02c33853fc46fb16e)

File: controllers/hvpa_controller.go [hunk 3]
```diff
@@ -274,15 +274,32 @@ func (r *HvpaReconciler) deleteHvpaFinalizers(hvpa *autoscalingv1alpha1.Hvpa) {
...
+	if ownerRef := metadata.GetOwnerReferences(); len(ownerRef) != 0 {
+		// TODO: Could be done better as part of validation
+		return nil, fmt.Errorf("vpa template in hvpa object already has an owner reference")
+	}
+
+	if generateName := metadata.GetGenerateName(); len(generateName) != 0 {
+		log.V(3).Info("Warning", "Generate name provided in the vpa template will be ignored", generateName)
+	}
+
+	metadata.SetName(hvpa.Name + "-vpa")
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: Shouldn't we do a deep copy before this?

File: controllers/hvpa_controller.go [hunk 4]
```diff
@@ -340,11 +360,28 @@ func (r *HvpaReconciler) reconcileVpa(hvpa *autoscalingv1alpha1.Hvpa) (*vpa_api.
...
+	if ownerRef := metadata.GetOwnerReferences(); len(ownerRef) != 0 {
+		// TODO: Could be done better as part of validation
+		return nil, fmt.Errorf("hpa template in hvpa object already has an owner reference")
+	}
+
+	if generateName := metadata.GetGenerateName(); len(generateName) != 0 {
+		log.V(3).Info("Warning", "Generate name provided in the hpa template will be ignored", generateName)
+	}
+
+	metadata.SetName(hvpa.Name + "-hpa")
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: Same as [above](https://github.com/gardener/hvpa-controller/pull/26/files#diff-6354ac486a904d10093c497f09f83a3eR295).

**Review by amshuman-kr — COMMENTED** [review]
Don't we need to deep copy the `ObjectMeta` from the templates?

**Review by amshuman-kr — APPROVED** [review]
LGTM

**Pull request merged**